channels/rdpsnd/client/rdpsnd_main.c

```diff
@@ -1273,47 +1000,43 @@ static VOID VCAPITYPE rdpsnd_virtual_channel_open_event_ex(LPVOID lpUserParam, D
 
 static DWORD WINAPI rdpsnd_virtual_channel_client_thread(LPVOID arg)
 {
-	wStream* data;
-	wMessage message;
+	BOOL running = TRUE;
 	rdpsndPlugin* rdpsnd = (rdpsndPlugin*) arg;
-	UINT error;
+	DWORD error = CHANNEL_RC_OK;
+	HANDLE events[2];
 
 	if ((error = rdpsnd_process_connect(rdpsnd)))
 	{
 		WLog_ERR(TAG, "error connecting sound channel");
 		goto out;
 	}
 
-	while (1)
+	events[1] = rdpsnd->stopEvent;
+	events[0] = Queue_Event(rdpsnd->queue);
+
+	do
 	{
-		if (!MessageQueue_Wait(rdpsnd->MsgPipe->In))
-		{
-			WLog_ERR(TAG, "MessageQueue_Wait failed!");
-			error = ERROR_INTERNAL_ERROR;
-			break;
-		}
+		const DWORD status = WaitForMultipleObjects(ARRAYSIZE(events), events, FALSE, INFINITE);
 
-		if (!MessageQueue_Peek(rdpsnd->MsgPipe->In, &message, TRUE))
+		switch (status)
 		{
-			WLog_ERR(TAG, "MessageQueue_Peek failed!");
-			error = ERROR_INTERNAL_ERROR;
-			break;
-		}
-
-		if (message.id == WMQ_QUIT)
-			break;
+			case WAIT_OBJECT_0:
+				{
+					wStream* s = Queue_Dequeue(rdpsnd->queue);
```

Comment:
Here if multiple streams are enqueued, only one is popped, this may lead to a dead lock or at least a delay in the sample rendering.

Reply:
@hardening No, the queue event is only reset if the queue is empty. Therefore the `WaitForMultipleObjects` will fire as long as there is an element in the queue.

Reply:
ok nice then